### PR TITLE
[web][docs] Add supports for push notification in Expo for Web

### PR DESCRIPTION
**Pull request by hesyifei**
# Why

Major web browsers (Chrome and Firefox) now all support standardized web push notifications (https://developer.mozilla.org/en-US/docs/Web/API/Push_API). With this PR, developers can use their existing code to also support push notifications in expo-web.

# How

- Added `Notifications.getExpoPushTokenAsync`, `Notifications.getDevicePushTokenAsync`, and `Notifications.addListener` supports for Expo for Web.
- Added docs "Using VAPID for Web Push Notifications".
- Add `notification.vapidPublicKey` in `app.json` and also supports `notification.icon` in `app.json`.
- Add supports for `icon`, `webPath`, `vapidSubject`, `_tag`, and `_renotify` in individual push messages.

This PR is complemented by https://github.com/expo/expo-cli/pull/844 and https://github.com/expo/universe/pull/3610.

# Test Plan

1. Clone expo-cli branch https://github.com/expo/expo-cli/pull/844 as well as the server branch https://github.com/expo/universe/pull/3610.
2. Upload/generate the VAPID keys and VAPID subject following the instruction in `docs/pages/versions/unversioned/guides/using-vapid.md`. Also add `owner` and `notification.vapidPublicKey` in `app.json`.
3. `<path>/expo-cli/node_modules/.bin/expo-cli start`
4. Open http://localhost:19006/ in the browser. (To use the IP address, follow the instruction here: expo/expo#4515)
5. Call `Notifications.getExpoPushTokenAsync()` in your client app and you should get a Expo token. (`native-component-list` should also work as expected without any code changes too.)
6. Send push notifications as usual - e.g. use https://expo.io/dashboard/notifications.

### CODE REVIEW

**Review by esamelson — CHANGES_REQUESTED** [review]
super cool, excited for this 😎

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 1]
```diff
@@ -113,4 +118,92 @@ export default {
...
+      body: JSON.stringify(tokenArguments),
+    }).then(response => response.json());
+
+    // TODO: Error handling
+    return response.data.expoPushToken;
+  },
+
+  async getDevicePushTokenAsync(): Promise<{ type: string; data: Object }> {
+    const data = await _subscribeUserToPushAsync();
+    return { type: 'web', data: data };
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]:
```suggestion
    return { type: Platform.OS, data: data };
```
In the future other platforms might fallback to web.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 2]
```diff
@@ -113,4 +118,92 @@ export default {
...
+
+async function _subscribeUserToPushAsync(): Promise<Object> {
+  if (!Constants.manifest.notification || !Constants.manifest.notification.vapidPublicKey) {
+    throw new CodedError(
+      'ERR_WEB_PUSH_NOTIFICATIONS_MISSING_CONFIG',
+      'You must provide `notification.vapidPublicKey` in `app.json` to use push notifications on web. Read more here: https://docs.expo.io/versions/latest/guides/using-vapid/.'
+    );
+  }
+  if (!('serviceWorker' in navigator)) {
+    throw new Error(
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: `isSupported` or just simply hide it behind the permissions.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 3]
```diff
@@ -113,4 +118,92 @@ export default {
...
+      `Current browser does not support Service Worker and hence do not support web push notifications`
+    );
+  }
+
+  const registration = await navigator.serviceWorker.register('/custom-service-worker.js');
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: Where does this come from? Is `custom-service-worker.js` the best name for an internal SW

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 4]
```diff
@@ -113,4 +118,92 @@ export default {
...
+  const pushSubscriptionJson = pushSubscription.toJSON();
+
+  const subscriptionObject = {
+    endpoint: pushSubscriptionJson.endpoint,
+    keys: {
+      p256dh: pushSubscriptionJson.keys!.p256dh,
+      auth: pushSubscriptionJson.keys!.auth,
+    },
+  };
+  console.log('subscriptionObject: ', JSON.stringify(subscriptionObject));
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: Remove

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 3 above)
Comment (hesyifei) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: I explained about that here: https://github.com/expo/expo-cli/pull/844#issuecomment-512917387. Though I do agree the name is not ideal😅.

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 5]
```diff
@@ -113,4 +118,92 @@ export default {
...
+
+  if (registration.active) {
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: May want to send something back for the dev to know if the SW isn't registered

File: packages/expo/src/Notifications/ExponentNotifications.web.ts [hunk 6]
```diff
@@ -113,4 +118,92 @@ export default {
   async cancelAllScheduledNotificationsAsync(): Promise<void> {
     this.dismissNotification();
   },
+
+  async getExponentPushTokenAsync(): Promise<string> {
+    if (!Constants.manifest.owner || !Constants.manifest.slug) {
+      throw new CodedError(
+        'ERR_WEB_PUSH_NOTIFICATIONS_MISSING_CONFIG',
```
Comment (EvanBacon) [inline packages/expo/src/Notifications/ExponentNotifications.web.ts]: Add an Error Codes section under the method that invokes this in the docs, like https://docs.expo.io/versions/latest/sdk/screen-orientation/#screenorientationlockasyncorientationlock. Also add a table of error codes to the bottom of the docs like: https://docs.expo.io/versions/latest/sdk/screen-orientation/#error-codes-2

File: packages/expo/src/Notifications/ExponentNotifications.web.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (hesyifei) [reply packages/expo/src/Notifications/ExponentNotifications.web.ts]: I think it is already handled in `expo-permission` (`return { status: PermissionStatus.UNDETERMINED, expires: 'never' };`) so we don't need to do any additional checking (the user should already checked `status === 'granted'`). I'll just use `guardPermission()` here.

**Review by esamelson — APPROVED** [review]
looks great to me, thanks!

**Review by xcarpentier — APPROVED** [review]

**Comment by xcarpentier** [comment]
Why this documentation is not available?
https://docs.expo.io/versions/latest/guides/using-vapid/

**Comment by ide** [comment]
It is not deployed yet. When we have an update available, we will update the docs.

**Comment by ide** [comment]
It is not deployed yet. When we have an update available, we will update the docs.

**Comment by djbuch** [comment]
Hello,
the documentation is available now, but I'm still getting an error when trying to generate the VAPID 

```
Reading project configuration...
Logging in...
Generating and setting VAPID keys on Expo's servers...
There is no PUT API endpoint at this URL.
```
Any thoughts ?

Thanks

**Comment by kopax** [comment]
Any update on this? I can't find how to generate the VAPID key on the documentation. Is this working in expo SDK 36? I was not able to upgrade to 37 and didn't even tried with expo sdk 38.

**Pull request merged**